[ko] Update outdated content in /concepts/scheduling-eviction/assign-pod-node (#53952)
kubernetes/website: content/ko/docs/concepts/scheduling-eviction/assign-pod-node.md

Reviewer: ```suggestion
* [테인트(Taints)와 톨러레이션(Tolerations)](/docs/concepts/scheduling-eviction/taint-and-toleration/)에 대해 더 읽어본다.
* [노드 어피니티](https://git.k8s.io/design-proposals-archive/scheduling/nodeaffinity.md)와
  [파드 간 어피니티/안티-어피니티](https://git.k8s.io/design-proposals-archive/scheduling/podaffinity.md)에 대한 디자인 문서를 읽어본다.
* [토폴로지 매니저](/docs/tasks/administer-cluster/topology-manager/)가 
  노드 수준 리소스 할당 결정에 어떻게 관여하는지 알아본다.
* [노드셀렉터(nodeSelector)](/docs/tasks/configure-pod-container/assign-pods-nodes/)를 어떻게 사용하는지 알아본다.
* [어피니티/안티-어피니티](/docs/tasks/configure-pod-container/assign-pods-nodes-using-node-affinity/)를 어떻게 사용하는지 알아본다.
```
Since the documentation now automatically redirects users to the version in their preferred language, it is no longer necessary to explicitly include the /ko prefix in the links.
And also I modified it to fit Korean grammar. `파드간` -> `파드 간`.

### 파드간 어피니티와 안티-어피니티 {#inter-pod-affinity-and-anti-affinity}

파드간 어피니티와 안티-어피니티를 사용하여, 

> 노드 레이블 대신, 각 노드에 이미 실행 중인 다른 파드 의 레이블을 기반으로 

Reviewer 2: ```suggestion
노드 레이블 대신, 각 노드에 이미 실행 중인 다른 파드의 레이블을 기반으로 
```

> 노드 레이블 대신, 각 노드에 이미 실행 중인 다른 **파드** 의 레이블을 기반으로 

Reviewer 3: Original sentence does not use `**` for pods.

노드 레이블 대신, 각 노드에 이미 실행 중인 다른 **파드** 의 레이블을 기반으로 
파드가 스케줄링될 노드를 제한할 수 있다.

> #### 파드간 어피니티 및 안티-어피니티 종류 {#types-of-inter-pod-affinity-and-anti-affinity}

Reviewer: ```suggestion
#### 파드 간 어피니티 및 안티-어피니티 종류 {#types-of-inter-pod-affinity-and-anti-affinity}
```

파드간 안티-어피니티를 사용하려면, 
파드 스펙에 `affinity.podAntiAffinity` 필드를 사용한다.

#### 스케줄링 동작 {#scheduling-behavior}

쿠버네티스 스케줄러가 새 파드를 스케줄링 할 때는 현재 클러스터 상태에서 파드의 어피니티와 안티-어피니티 규칙을 평가한다: 

1. 하드 제약 조건 (노드 필터링)
   - `podAffinity.requiredDuringSchedulingIgnoredDuringExecution`와 `podAntiAffinity.requiredDuringSchedulingIgnoredDuringExecution`:
     - 스케줄러가 기존 파드들 기준으로 어피니티와 안티-어피니티 규칙을 만족하는 노드들에만 새 파드를 배치한다.

2. 소프트 제약 조건 (스코어링)
   - `podAffinity.preferredDuringSchedulingIgnoredDuringExecution`와 `podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution`:
     - 스케줄러가 선호되는 어피니티와 안티-어피니티 규칙에 노드들이 얼마나 잘 충족하는지에 따라 점수를 지정하며 파드 배치를 최적화한다.

3. 무시되는 필드:
   - 기존 파드들의 `podAffinity.preferredDuringSchedulingIgnoredDuringExecution`:
     - 이러한 선호되는 어피니티 규칙들은 새 파드 스케줄링 결정에서 고려되지 않는다.
   - 기존 파드들의 `podAntiAffinity.preferredDuringSchedulingIgnoredDuringExecution`:

>      - 마찬가지로, 기존 파드들의 선호되는 안티-어피니티 규칙들은 스케줄링에 고려되지 않는다.

Reviewer: ```suggestion
     - 마찬가지로, 기존 파드에 설정된 선호 안티-어피니티 규칙도 스케줄링 시 무시된다.
```
I modified it to fit Korean grammar.

파드 어피니티 규칙은 "하드" `requiredDuringSchedulingIgnoredDuringExecution`을, 
안티-어피니티 규칙은 "소프트" `preferredDuringSchedulingIgnoredDuringExecution`을 사용한다.

> 위의 어피니티 규칙은 기존 파드에 `security=S1` 레이블이 있는 

Reviewer: ```suggestion
해당 어피니티 규칙은 `security=S1` 레이블이 지정된 파드가 이미 실행 중인 특정  

```

[존](/docs/concepts/scheduling-eviction/topology-spread-constraints/)의 

> 노드에만 파드를 스케줄링하도록 스케줄러에 지시한다.

Reviewer: ```suggestion
노드에만 예시 파드를 배치할 수 있도록 스케줄러에 지시한다.
```

예를 들어, `topology.kubernetes.io/zone=V` 레이블이 있는 노드들로 
구성된 존 "Zone V"이 있다고 가정해보면, Zone V에 `security=S1` 레이블이 있는 
하나 이상의 기존 파드가 있는 한 스케줄러는 Zone V 내의 어느 노드에나 파드를 
배치할 수 있다. 반대로, `security=S1` 레이블이 있는 파드가 Zone V 내에 없다면, 

> 스케줄러는 그 존의 어떤 노드에도 위 예시의 파드를 배치하지 않을 것이다.

Reviewer: ```suggestion
구성된 "Zone V"라는 존이 있다고 가정해보면, Zone V 내에 `security=S1` 레이블이 있는 
파드가 최소 하나 이상 존재한다면 스케줄러는 해당 존 내의 어느 노드에나 파드를 
할당할 수 있다. 반대로, `security=S1` 레이블이 있는 파드가 Zone V 내에 없다면, 
스케줄러는 해당 존의 어떤 노드에도 예시 파드를 할당하지 않는다.
```

> 위의 안티-어피니티 규칙은 `security=S2` 레이블이 있는 기존 파드가 하나 이상이 

Reviewer: ```suggestion
해당 안티-어피니티 규칙은 `security=S2` 레이블이 지정된 파드가 
```

쿠버네티스에는 파드 어피니티 또는 안티-어피니티를 위한 선택적 `matchLabelKeys` 필드가
존재한다. 이 필드는 파드 (안티)어피니티를 충족할 때, 들어오는 파드의 레이블과
일치해야 하는 레이블 키를 지정한다.

레이블 키는 파드 레이블 값들을 조회할 때 사용된다. 이러한 키-값 레이블들은 
`labelSelector` 필드로 정의된 매치 제약과 결합된다 (`AND` 연산자 사용). 
결합된 필터링은 파드 (안티)어피니티 계산에 포함될 기존 파드들을 선택한다.

{{< caution >}}
`matchLabelKeys`를 파드에서 직접 업데이트될 수 있는 레이블과 함께 사용하는 것은 권장되지 않는다.
`matchLabelKeys`에 **직접** (즉, 디플로이먼트를 통하지 않고) 명시된 파드의 레이블을 수정할 때에도
API 서버(kube-apiserver)는 병합된 `labelSelector`의 레이블 업데이트를 반영하지 않는다.
{{< /caution >}}

흔한 유스케이스에는 `matchLabelKeys`를 `pod-template-hash`와 함께 사용하는 
것이 있다 (디플로이먼트의 일부로 관리되는 파드에서 설정되며, 각 리비전의 값이 고유함). 
`matchLabelKeys`에서 `pod-template-hash`를 사용하면 들어오는 파드와 동일한 리비전에
속하는 파드들을 타겟팅함으로써 롤링 업그레이드시 어피니티가 깨지지 않을 수 있다.

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: application-server
...
spec:
  template:
    spec:
      affinity:
        podAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: app
                operator: In
                values:
                - database
            topologyKey: topology.kubernetes.io/zone
            # 주어진 롤아웃의 파드만이 파드 어피니티를 계산하는 데 고려된다.

>             # 디플로이먼트를 업데이트할 시, 교체되는 파드들은 자신들의 어피니티 규칙들을 따른다

Reviewer 3: Period (.) is missing: `따른다` -> `따른다.`

            # (새로운 파드 템플릿에 정의된 규칙이 있다면)
            matchLabelKeys:
            - pod-template-hash
```

#### mismatchLabelKeys {#mismatchlabelkeys}

{{< feature-state feature_gate_name="MatchLabelKeysInPodAffinity" >}}

{{< note >}}
<!-- UPDATE THIS WHEN PROMOTING TO STABLE -->
`mismatchLabelKeys`는 베타 단계 필드이며 쿠버네티스 {{< skew currentVersion >}}에서 
기본으로 활성화되어 있다.
이를 비활성화 하려면, `MatchLabelKeysInPodAffinity` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)에서 
명시적으로 비활성화 해야 한다.
{{< /note >}}

쿠버네티스에는 파드 어피니티 또는 안티-어피니티를 위한 선택적 `mismatchLabelKeys` 필드가
존재한다. 이 필드는 파드 안티-어피니티를 충족할 때 들어오는 파드의 레이블과
불일치해야 하는 레이블 키들을 지정한다.

{{< caution >}}
`mismatchLabelKeys`를 파드에서 직접 업데이트될 수 있는 레이블과 함께 사용하는 것은 권장되지 않는다.
`mismatchLabelKeys`에 **직접** (즉, 디플로이먼트를 통하지 않고) 명시된 파드의 레이블을 수정할 때에도
API 서버(kube-apiserver)는 병합된 `labelSelector`의 레이블 업데이트를 반영하지 않는다.
{{< /caution >}}

> 예시로 파드들이 동일한 테넌트(tenant) 또는 팀이 스케줄되어 있는 토폴로지 도메인에 (노드, 존 등) 배치되도록 할 수 있다.

Reviewer: ```suggestion
한 가지 예로 파드들이 동일한 테넌트(tenant) 또는 팀이 스케줄되어 있는 토폴로지 도메인에 (노드, 존 등) 배치되도록 할 수 있다.
```

I modified it to fit Korean grammar.

즉, 서로 다른 테넌트의 파드들이 같은 토폴로지 도메인에서 동시에 실행되는 것을 막을 수 있다.

```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    # 모든 관련 파드에 "tenant" 레이블이 설정되어 있다고 가정한다
    tenant: tenant-a
...
spec:
  affinity:
    podAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:

>       # 이 테넌트와 연관된 파드들이 알맞는 노드 풀에 배치되도록 한다

Reviewer: ```suggestion
      # 이 테넌트와 연관된 파드들이 올바른 노드 풀에 배치되도록 보장한다
```

I modified it to fit Korean grammar.

      - matchLabelKeys:
          - tenant
        labelSelector: {}
        topologyKey: node-pool
    podAntiAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      # 이 테넌트와 연관된 파드들이 다른 테넌트가 사용하는 노드에 스케줄되지 않도록 보장한다
      - mismatchLabelKeys:
        - tenant # 이 파드의 "tenant" 레이블 값이 무엇이든지,
                 # 다른 테넌트에 속한 파드가 실행되고 있는 풀의 노드에는 스케줄링 되지

>                  # 않도록 한다

Reviewer 3: Period (.) is missing: `한다` -> `한다.`

`nodeName` 을 사용해서 노드를 선택할 때의 몇 가지 제한은 다음과 같다.

- 만약 명명된 노드가 없으면, 파드가 실행되지 않고

>   따라서 자동으로 삭제될 수 있다.

Reviewer: ```suggestion
- 만약 명명된 노드가 없으면, 파드가 실행되지 않으며,
  경우에 따라 자동으로 삭제될 수도 있다.
```

I modified it to fit Korean grammar.

- 만약 명명된 노드에 파드를 수용할 수 있는
  리소스가 없는 경우 파드가 실패하며, 그 이유는 

>   예를 들어 OutOfmemory 또는 OutOfcpu와 같이 왜 살패했는지를 나타낸다.

Reviewer 2: ```suggestion
  예를 들어 OutOfmemory 또는 OutOfcpu와 같이 왜 실패했는지를 나타낸다.
```

위 파드는 `kube-01` 노드에서만 실행될 것이다.

## nominatedNodeName {#nominatednodename}

{{< feature-state feature_gate_name="NominatedNodeNameForExpectation" >}}

`nominatedNodeName`은 외부 컴포넌트가 pending 파드를 위한 노드를 지명하는 데 사용될 수 있다.

> 이 지명은 최선의 노력이며, 스케줄러가 해당 파드가 지명된 노드로 갈 수 없다고 판단하면 무시될 수 있다.

Reviewer: ```suggestion
이 지명은 최선의 노력이며, 스케줄러가 해당 파드가 지명된 노드에 배치될 수 없다고 판단하면 무시될 수 있다.
```

또한 이 필드는 스케줄러에 의해 (덮어)쓸 수 있다:
- 스케줄러가 선점(preemption)을 통해 지명할 노드를 찾는 경우 .
- 스케줄러가 파드가 갈 위치를 결정하고 바인딩 사이클로 이동하는 경우.
  - 이 경우 `nominatedNodeName`은 파드가 `WaitOnPermit`이나 `PreBind` 익스텐션(extension) 포인트를 거쳐야 할 때만 설정된다.

> 다음은 `nominatedNodeName` 필드를 사용한 파드 상태이다:

Reviewer: ```suggestion
다음은 `nominatedNodeName` 필드를 사용한 파드 상태의 예시이다.
```

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: nginx
...
status:
  nominatedNodeName: kube-01
```

## 파드 토폴로지 분배 제약 조건 {#pod-topology-spread-constraints}

_토폴로지 분배 제약 조건_을 사용하여 지역(regions), 영역(zones), 노드 또는 
사용자가 정의한 다른 토폴로지 도메인과 같은 장애 도메인 사이에서 {{< glossary_tooltip text="파드" term_id="Pod" >}}가 
클러스터 전체에 분산되는 방식을 제어할 수 있다. 성능, 예상 가용성 또는 전체 활용도를 

> 개선하기 위해 이 작업을 수행할 수 있다.

Reviewer: ```suggestion
_토폴로지 분배 제약 조건_을 사용하여 리전(regions), 영역(zones), 노드 또는 
사용자가 정의한 다른 토폴로지 도메인과 같은 장애 도메인(failer-domains) 에 {{< glossary_tooltip text="파드" term_id="Pod" >}}가 
클러스터 전체에 분산되는 방식을 제어할 수 있다. 성능 향상, 가용성 확보 또는 전반적인 자원 활용도 개선을 
위해 이 작업을 수행할 수 있다.
```

[파드 토폴로지 분배 제약 조건](/ko/docs/concepts/scheduling-eviction/topology-spread-constraints/)에서 
작동 방식에 대해 더 자세히 알아볼 수 있다.

## 파드 토폴로지 레이블 {#pod-topology-labels}

{{< feature-state feature_gate_name="PodTopologyLabelsAdmission" >}}

> 파드는 배치된 노드에 토폴로지 레이블(`topology.kubernetes.io/zone`과 `topology.kubernetes.io/region`)이 있을 경우 이를 물려받는다. 이 레이블들은 다운워드(Downward) API를 통해 워크로드에 노드 토폴로지 인식을 제공하는 데 사용될 수 있다.

Reviewer: ```suggestion
파드는 할당된 노드에 토폴로지 레이블(`topology.kubernetes.io/zone`과 `topology.kubernetes.io/region`)이 존재할 경우 이를 해당 노드로부터 물려받는다. 이 레이블들은 다운워드(Downward) API를 통해 워크로드가 노드 토폴로지를 인식할 수 있도록 제공하는 데 활용될 수 있다.
```

다음은 존과 영역(region)에 다운워드 API를 사용하는 파드의 예시이다:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: pod-with-topology-labels
spec:
  containers:
    - name: app
      image: alpine
      command: ["sh", "-c", "env"]
      env:
        - name: MY_ZONE
          valueFrom:
            fieldRef:
              fieldPath: metadata.labels['topology.kubernetes.io/zone']
        - name: MY_REGION
          valueFrom:
            fieldRef:
              fieldPath: metadata.labels['topology.kubernetes.io/region']
```

## 연산자 {#operators}

다음은 위에서 언급된 `nodeAffinity`와 `podAffinity`의 `operator` 필드에서 사용할 수 있는 모든 논리 연산자들이다.

|    연산자    |    동작     |
| :------------: | :-------------: |
| `In` | 레이블 값이 주어진 문자열 집합에 존재한다 |
|   `NotIn`   | 레이블 값이 주어진 문자열 집합에 존재하지 않는다 |
| `Exists` | 레이블 키가 오브젝트에 존재한다 |
| `DoesNotExist` | 레이블 키가 오브젝트에 존재하지 않는다 |

다음 연산자들은 `nodeAffinity`에만 사용될 수 있다.

|    Operator    |    Behavior    |
| :------------: | :-------------: |
| `Gt` | 필드의 값이 정수로 해석되고, 이 셀렉터에 의해 명명된 레이블의 값에서 해석된 정수 값보다 작다 |

> | `Lt` | 필드의 값이 정수로 해석되고, 이 셀렉터에 의해 명명된 레이블의 값에서 해석된 정수 값보다 크다 |

Reviewer: ```suggestion
| `Gt` | 필드 값을 정수로 해석하며, 그 값이 셀렉터에 지정된 레이블 값을 해석한 정수보다 작다 |
| `Lt` | 필드 값을 정수로 해석하며, 그 값이 셀렉터에 지정된 레이블 값을 해석한 정수보다 크다 |
```
I modified it to fit Korean grammar.

파드는 할당된 노드에 토폴로지 레이블(`topology.kubernetes.io/zone`과 `topology.kubernetes.io/region`)이 존재할 경우 이를 해당 노드로부터 물려받는다. 이 레이블들은 다운워드(Downward) API를 통해 워크로드가 노드 토폴로지를 인식할 수 있도록 제공하는 데 활용될 수 있다.

다음은 영역과 리전에 다운워드 API를 사용하는 파드의 예시이다.
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: pod-with-topology-labels
spec:
  containers:
    - name: app
      image: alpine
      command: ["sh", "-c", "env"]
      env:
        - name: MY_ZONE
          valueFrom:
            fieldRef:
              fieldPath: metadata.labels['topology.kubernetes.io/zone']
        - name: MY_REGION
          valueFrom:
            fieldRef:
              fieldPath: metadata.labels['topology.kubernetes.io/region']
```

## 연산자 {#operators}

다음은 위에서 언급된 `nodeAffinity`와 `podAffinity`의 `operator` 필드에서 사용할 수 있는 모든 논리 연산자이다.

|    연산자    |    동작     |
| :------------: | :-------------: |
| `In` | 레이블 값이 주어진 문자열 집합에 존재한다 |
|   `NotIn`   | 레이블 값이 주어진 문자열 집합에 존재하지 않는다 |
| `Exists` | 레이블 키가 오브젝트에 존재한다 |
| `DoesNotExist` | 레이블 키가 오브젝트에 존재하지 않는다 |

다음 연산자들은 `nodeAffinity`에만 사용될 수 있다.

|    Operator    |    Behavior    |
| :------------: | :-------------: |
| `Gt` | 필드의 값이 정수로 해석되고, 이 셀렉터에 의해 명명된 레이블의 값에서 해석된 정수 값이 이 값보다 크다 |

> | `Lt` | 필드의 값이 정수로 해석되고, 이 셀렉터에 의해 명명된 레이블의 값에서 해석된 정수 값이 이 값보다 작다 |

Reviewer 2: @thdxg @developowl 

죄송합니다 제가 인용으로 원문을 가져올 때 잘못된 문장으로 가져왔습니다. 현재는 수정하여 지금 나와있는 문장이 정확한 문장입니다.

수정된 문장으로 고쳐보면, 

> | Gt | The field value will be parsed as an integer, and the integer that results from parsing the value of a label named by this selector is greater than this integer |

필드 값은 정수로 파싱되며, 이 셀렉터가 지정한 라벨 값에서 파싱된 정수가 필드 값 정수보다 크다.

> | Lt | The field value will be parsed as an integer, and the integer that results from parsing the value of a label named by this selector is less than this integer |

필드 값은 정수로 파싱되며, 이 셀렉터가 지정한 라벨 값에서 파싱된 정수가 필드 값 정수보다 작다.

위와 같이 될 것 같은데 어떠신지요?

두 값을 비교하는 문장인데 큰 혼란을 드려 죄송합니다🙇🏻‍♂️

추가로 저 역시 다소 번역투의 느낌이 나지만 원문을 최대한 살리기 위해 문장 구조를 생략하지 않는 방향이 좋다고 생각합니다.

다시 한 번 확인 부탁드립니다.

> | `Lt` | 필드의 값이 정수로 해석되고, 이 셀렉터에 의해 명명된 레이블의 값에서 해석된 정수 값이 이 값보다 작다 |

Author: 저도 @wonyongg 께서 제안해주신 번역이 좋은 것 같습니다! 맞춰서 수정하도록 하겠습니다!

* [토폴로지 매니저](/docs/tasks/administer-cluster/topology-manager/)가 
  노드 수준 리소스 할당 결정에 어떻게 관여하는지 알아본다.

> * [노드셀렉터(nodeSelector)](/ko/docs/tasks/configure-pod-container/assign-pods-nodes/)를 어떻게 사용하는지 알아본다.

Reviewer: Please delete this section.
